docs/resources/project_protected_environment.md

```diff
@@ -0,0 +1,63 @@
+# gitlab\_project\_protected\_environment
+
+This resource allows you to create and manage protected environments for your
+GitLab projects. For further information on project protected environments, 
+consult the [GitLab documentation](https://docs.gitlab.com/ee/ci/environments/protected_environments.html)
+and the [API documentation](https://docs.gitlab.com/ee/api/protected_environments.html).
+
+-> This feature requires a GitLab Premium account or above.
+
+## Example Usage
+
+```hcl
+resource "gitlab_project_environment" "example-one" {
+  project      = 5
+  name         = "production/web"
+  external_url = "https://example.com"
+}
+resource "gitlab_project_protected_environment" "example-one" {
+  # depends_on ensures the environment exists before protecting it; if
+  # environment is defined in a .gitlab-ci.yml file or otherwise, this is not
+  # required
+  depends_on  = [gitlab_project_environment.example-one]
+  project     = 5
+  environment = "production/*"
+  deploy_access_levels {
+    access_level = "maintainer"
+  }
+  deploy_access_levels {
+    user_id = 1
+  }
+  deploy_access_levels {
+    group_id = 2
+  }
+}
+```
+
+## Argument Reference
+
+The following arguments are supported:
+
+* `project` - (Required, string) The name or id of the project to add the project protected environments.
+
+* `environment` - (Required) The name of the project protected environment, or a wildcard expression for which environments to protect.
+
+* `deploy_access_levels` - (Required) One or more `deploy_access_levels` blocks as defined below.
+
+---
+
+A `deploy_access_levels` block supports the following:
+
+* `access_level` - (Optional) The access level which is allowed to deploy to the protected environment. Valid values are `developer` or `maintainer`. Mutually exclusive with `user_id` and `group_id`.
```

Comment:
> Valid values are `developer` or `maintainer`.

Sorry if I'm missing it, where is this being validated?

Reply:
Only server-side, I don't think I've added validation in the schema.

Reply:
I guess the note on the API docs means I can use StringInSlice for validating `developer`, `maintainer` or `admin`: https://docs.gitlab.com/ee/api/protected_environments.html#valid-access-levels

Reply:
Turns out I did validate it here, though it would also accept "reporter", etc.

https://github.com/gitlabhq/terraform-provider-gitlab/blob/2cc27a4e0f881ad64f590f58df7290f98c69f575/gitlab/resource_gitlab_project_protected_environment.go#L45-L50

Would you like me to change it to strictly allow `developer`, `maintainer` or `admin`? The UI only provides the first two as options.

Reply:
@sirlatrom Yeah, I think it would be good to make the `ValidateFunc` match the documentation 👍

Reply:
That should be done now, and I've rebased on master, too.